graylog2-server/src/main/java/org/graylog2/database/dbcatalog/DbEntitiesScanner.java

```diff
@@ -16,88 +16,32 @@
     }
 
     @Override
     public DbEntitiesCatalog get() {
-        final Stopwatch stopwatch = Stopwatch.createStarted();
-        final DbEntitiesCatalog catalog = scan(packagesToScan, packagesToExclude, chainingClassLoader);
-        stopwatch.stop();
-        LOG.info("{} entities have been scanned and added to DB Entity Catalog, it took {}", catalog.size(), stopwatch);
-        return catalog;
-    }
-
-    public DbEntitiesCatalog scan(final String[] packagesToScan,
-                                  final String[] packagesToExclude,
-                                  final ChainingClassLoader chainingClassLoader) {
-        List<DbEntityCatalogEntry> dbEntities = new LinkedList<>();
-        ClassGraph classGraph = new ClassGraph()
-                .enableAnnotationInfo()
-                .acceptPackages(packagesToScan)
-                .rejectPackages(packagesToExclude)
-                .filterClasspathElements(classpathElementPathStr -> classpathElementPathStr.contains("graylog"))
-                .disableRuntimeInvisibleAnnotations();
-
-        if (chainingClassLoader != null) {
-            //Unfortunately, ClassGraph does not work correctly if provided with ChainingClassLoader as a whole
-            //You have to manually add all class loaders from ChainingClassLoader
-            for (ClassLoader cl : chainingClassLoader.getClassLoaders()) {
-                classGraph = classGraph.addClassLoader(cl);
-            }
-        }
-
-        try (ScanResult scanResult = classGraph.scan()) {
-            final String annotationName = DbEntity.class.getCanonicalName();
-            ClassInfoList classInfoList = scanResult.getClassesWithAnnotation(annotationName);
-            for (ClassInfo classInfo : classInfoList) {
-                final var annotations = classInfo.getAnnotationInfoRepeatable(annotationName);
-                for (final var annotationInfo : annotations) {
-                    AnnotationParameterValueList paramVals = annotationInfo.getParameterValues();
-                    dbEntities.add(new DbEntityCatalogEntry(
-                            paramVals.get("collection").getValue().toString(),
-                            paramVals.get("titleField").getValue().toString(),
-                            classInfo.loadClass(),
-                            paramVals.get("readPermission").getValue().toString()
-                    ));
-                }
-            }
-
-            return new DbEntitiesCatalog(dbEntities);
-        }
+        final var catalogEntries = entityClasses.stream()
+                .flatMap(clazz ->
+                        Arrays.stream(clazz.getAnnotationsByType(DbEntity.class))
+                                .map(a ->
+                                        new DbEntityCatalogEntry(a.collection(), a.titleField(), clazz,
+                                                a.readPermission()))
+                )
+                .toList();
+        return new DbEntitiesCatalog(catalogEntries);
```

Comment:
How about making the class a singleton and memoize the `catalogEntries` value? The classes don't change at runtime.

Reply:
Because we are binding `DbEntitiesCatalog` as an eager singleton, we are doing the annotation processing during initialization anyway. I moved the catalog creation to the constructor. This way we don't have to add extra memoization logic.

Reply:
Thank you! :pray: